common/djangoapps/student/roles.py

```diff
@@ -83,15 +111,32 @@ def get_user_roles(cls, user):
 
 class RoleCache:
     """
-    A cache of the CourseAccessRoles held by a particular user
+    A cache of the CourseAccessRoles held by a particular user.
+    Internal data structures should be accessed by getter and setter methods;
+    don't use `_roles_by_course_id` or `_roles` directly.
+    _roles_by_course_id: This is the data structure as saved in the RequestCache.
+        It contains all roles for a user as a dict that's keyed by course_id.
+        The key ROLE_CACHE_UNGROUPED_COURSES_KEY is used for all roles
+        that are not associated with a course.
+    _roles: This is a set of all roles for a user, ungrouped. It's used for some types of
+        lookups and collected from _roles_by_course_id on initialization
+        so that it doesn't need to be recalculated.
+
     """
     def __init__(self, user):
         try:
-            self._roles = BulkRoleCache.get_user_roles(user)
+            self._roles_by_course_id = BulkRoleCache.get_user_roles(user)
         except KeyError:
-            self._roles = set(
-                CourseAccessRole.objects.filter(user=user).all()
-            )
+            self._roles_by_course_id = {}
+            roles = CourseAccessRole.objects.filter(user=user).all()
+            for role in roles:
+                course_id = get_role_cache_course_key(role.course_id)
```

Comment:
This means that org level roles (where org is specified but course_id is null) will be dumped into the generic ungrouped cache key? Is that okay?

Reply:
That seems okay to me. I mean it was ungrouped before already, so it's not worse than before, and we're trying to improve performance for lookups for libraries that relate to a specific course, not for orgs. It could be an idea for a future improvement but it's not clear that that is necessary or has a benefit. Also it depends how many roles are in that group. If we can get to a strong assumption that the biggest mass of roles is already grouped because it's by course which should include libraries, then a user will normally not belong to thousands of different orgs, so the amount of ungrouped roles will be quite small und thus no big problem to iterate over.

Now a big caveat here is that I didn't get this to work with V2 libraries. It seems that V2 libraries don't have course keys with an html_id attribute and thus will not be groupable by course and land in the ungrouped set. So this performance improvement will not work for V2 libraries as-is, and I wanted to ask about it. However, you or Kyle pointed out that html_id is not the right attribute to look for the string representation, that it should be something like `str(course_key)`. If I understand that correctly and that works for everything including V2 libraries, that should resolve our problem here.

Reply:
> However, you or Kyle pointed out that html_id is not the right attribute to look for the string representation, that it should be something like str(course_key). If I understand that correctly and that works for everything including V2 libraries, that should resolve our problem here.

Yes, for any OpaqueKey instance (including CourseKeys, LibraryKeys, and UsageKeys), simply calling `str(...)` with the key is a reliable and surprise-free way of getting a string that you can use as a cache key, guaranteed to be unique from any other OpaqueKey instance.